rubyforgood/casa
# [#1071] Add filters for case contacts

### What github issue is this PR for, if any?
Resolves #1071 

### What changed, and why?
Volunteers can now filter and sort the case contacts for each of their cases.

#### Volunteers should be able to filter by
- [x] date ranges
- [ ] contact_type_group
- [x] case_contact_contact_type
- [x] contact_made (yes/no)
- [x] contact_medium
- [x] want_driving_reimbursement (yes/no)
- [x] Volunteers can sort all of the above by ascending and descending order.

I did not add a contact type group filter since the contact type filter already allows the user to choose by group.

#### Acceptance criteria
- [ ] A button labeled "Filtered by" displays in the new case_contacts table header viewable by volunteers (see #696)
- [ ] Clicking the button displays all "Filter by" options
- [x] Volunteer can filter by any combination of conditions above
- [x] It is easy for the volunteer to close the "Filter by" menu after making their selections

I didn't strictly follow the suggested design, because the new case contacts layout is not yet implemented, but I can always change the view if something is undesirable.

### How will this affect user permissions?
No permissions have been affected.

### How is this tested? (please write tests!) 💖💪
#### Automated Testing
I've added specs for the new scopes and some basic system specs to check if a filter is working.

#### Manual Testing
- Go to the case contacts page
- At the right side of the "Filter by" card, click on Show/Hide.
- Select the desired filters and click on "Filter".
- After clicking on the "Filter" button, the "Filter by" card should remain expanded
- Check if the filters and the sorting work as expected

### Screenshots please :)
#### Overview (Desktop)
![01-overview](https://user-images.githubusercontent.com/72531802/122577570-54ceaf00-d029-11eb-9ba3-51593b842cfd.png)

#### Filter section (Desktop)
![02-filters-expanded](https://user-images.githubusercontent.com/72531802/122577591-58facc80-d029-11eb-8012-af65441b5bb8.png)

#### Overview (Mobile)
![03-overview-mobile](https://user-images.githubusercontent.com/72531802/122577605-5bf5bd00-d029-11eb-9630-dcf310962ccd.png)

#### Filter section (Mobile)
![04-full-filter-mobile](https://user-images.githubusercontent.com/72531802/122577612-5ef0ad80-d029-11eb-9879-a0dc84a72724.png)
